File: axe/ast.py
from dataclasses import dataclass
from typing import Any


class ASTNode:
    pass


# ============================================================
# Program
# ============================================================

@dataclass
class Program(ASTNode):
    statements: list


# ============================================================
# Statements
# ============================================================

@dataclass
class ExpressionStatement(ASTNode):
    expression: Any


@dataclass
class VariableDeclaration(ASTNode):
    name: str
    initializer: Any
    constant: bool = False


@dataclass
class BlockStatement(ASTNode):
    statements: list


@dataclass
class IfStatement(ASTNode):
    condition: Any
    then_branch: Any
    else_branch: Any = None


@dataclass
class ReturnStatement(ASTNode):
    value: Any = None


@dataclass
class FunctionDeclaration(ASTNode):
    name: str
    parameters: list
    body: BlockStatement


# ============================================================
# Expressions
# ============================================================

@dataclass
class BinaryExpression(ASTNode):
    left: Any
    operator: str
    right: Any


@dataclass
class UnaryExpression(ASTNode):
    operator: Any
    operand: Any


@dataclass
class Literal(ASTNode):
    value: Any


@dataclass
class Identifier(ASTNode):
    name: str


@dataclass
class Assignment(ASTNode):
    name: str
    value: Any


@dataclass
class CallExpression(ASTNode):
    callee: Any
    arguments: list

File: src/ast.rs
use crate::token::TokenType;

#[derive(Debug)]
pub struct Program {
    pub statements: Vec<Statement>,
}

#[derive(Debug)]
pub enum Statement {
    VariableDeclaration {
        name: String,
        initializer: Expression,
        constant: bool,
    },

    Expression(Expression),

    Block {
        statements: Vec<Statement>,
    },

    If {
        condition: Expression,
        then_branch: Box<Statement>,
        else_branch: Option<Box<Statement>>,
    },

    Function {
        name: String,
        parameters: Vec<String>,
        body: Box<Statement>,
    },

    Return(Option<Expression>),
}

#[derive(Debug)]
pub enum Expression {
    Literal(LiteralValue),

    Identifier(String),

    Binary {
        left: Box<Expression>,
        operator: TokenType,
        right: Box<Expression>,
    },

    Unary {
        operator: TokenType,
        operand: Box<Expression>,
    },

    Assignment {
        name: String,
        value: Box<Expression>,
    },

    Call {
        callee: Box<Expression>,
        arguments: Vec<Expression>,
    },
}

#[derive(Debug)]
pub enum LiteralValue {
    Number(f64),
    String(String),
    Boolean(bool),
    Null,
}